### PR TITLE
ncm-grub: full overhaul

### LINKED ISSUE #232: ncm-grub: full rewrite needed!
ncm-grub needs a full rewrite to at least:
- use the getTree() API
- use CAF::Process to call `grubby` command
- Cleanup unnecessary updates of grub configuration (for example if kernel args defined in the configuration are already set).

### LINKED ISSUE #1000: ncm-grub: drop non-multiboot support
As part of #999, we could consider dropping support for grubby versions that do not support multiboot.

It would simplify the code a bit and remove a grub2-incompatible codepath.

On centos 5.11/6.5 and 7.2, grubby has multiboot support.

### PR DESCRIPTION
**Pull request by stdweird**
Fixes #232
Also fixes #1000 

Backwards incompatible becuase non-multiboot grubby support was dropped

### CODE REVIEW

**Comment by stdweird** [comment]
unittests will follow

File: ncm-grub/src/main/perl/grub.pm [hunk 1]
```diff
@@ -1,99 +1,116 @@
...
+=over
+
+=item grubby_args_options
+
+Given string C<args>, split and convert in grubby commandline options
+to add and/or remove the arguments.
+Arguments prefixed with '-' are scheduled for removal
+If C<multiboot> is true, generate multiboot commandline options
+
+Returns a list with options.
```
Comment (ned21) [inline ncm-grub/src/main/perl/grub.pm]: s/with/of/ ?

File: ncm-grub/src/main/perl/grub.pm [hunk 2]
```diff
@@ -1,99 +1,116 @@
...
-        chop($kernelargsremove);
-    }
+    my @options;
 
-    return ($kernelargsadd, $kernelargsremove);
+    my $mb = $multiboot ? 'mb' : '';
+    push(@options, "--${mb}args", join(" ", @add)) if @add;
+    push(@options, "--remove-${mb}args", join(" ", @remove)) if @remove;
+
+    $self->debug(1, "converted '$args' in options '@options'");
```
Comment (ned21) [inline ncm-grub/src/main/perl/grub.pm]: in => into

File: ncm-grub/src/main/perl/grub.pm [hunk 3]
```diff
@@ -124,496 +140,509 @@ sub password {
...
+by editing the grub conf via filehandle C<grub_fh>
+(a C<CAF::FileEditor> instance, which is not closed in this method).
 
-    my $grub_fh = CAF::FileEditor->new( $GRUB_CONF,
-        owner => "root",
-        group => "root",
-        mode  => 0400,
-        log   => $self );
+Returns undef on failure, the console kernel commandline option
+(or empty string if none is to be configured) on success.
```
Comment (ned21) [inline ncm-grub/src/main/perl/grub.pm]: Any particular reason why the return type is different to the password function? Why not return SUCCESS like it does?

File: ncm-grub/src/main/perl/grub.pm [hunk 4]
```diff
@@ -124,496 +140,509 @@ sub password {
...
 
-                    $grubbystring .= ($mbargsadd) ? " --mbargs=\"$mbargsadd\"" : "";
-                    $grubbystring .= ($mbargsremove) ? " --remove-mbargs=\"$mbargsremove\"" : "";
 
-                    $grubbystring .= " --add-kernel=\"$fullkernelpath\"";
+=item grubby
 
-                    $grubbystring .= ($kernelargsadd) ? " --args=\"$kernelargsadd\"" : "";
-                    $grubbystring .= ($kernelargsremove) ? " --args=\"$kernelargsremove\"" : "";
+Run C<grubby> with arraref C<args> via C<CAF::Proces> using the
```
Comment (ned21) [inline ncm-grub/src/main/perl/grub.pm]: Typo: missing s from CAF::Process

File: ncm-grub/src/main/perl/grub.pm (hunk 3 above)
Comment (stdweird) [inline ncm-grub/src/main/perl/grub.pm]: the method also generates the console kernel commandline option, not sure how to combine these 2 otherwise.

**Comment by stdweird** [comment]
@ned21 is now fully unittested; i opened #1002 and #1003 for followup (i think MS is one of the main users of serial and password settings?)

File: ncm-grub/src/main/pan/components/grub/schema.pan [hunk 5]
```diff
@@ -16,10 +16,25 @@ include 'quattor/schema';
...
+    "enabled" ? boolean
+    @{An --option used with the password line in grub.conf. This is typically
+      used to set the hashing algorithm for the password. "encrypted" means the
+      password can be hashed with (more secure than MD5) SHA-256 or SHA-512.
+      "md5" for an MD5 hashed password. Plaintext is not supported.}
     "option" : string with match (SELF, "^(md5|encrypted)$")
+    @{Mutually exclusive with the file option. A crypted password for grub.conf.}
     "password" ? string
-    "enabled" ? boolean
+    @{Mutually exclusive with the password option. The path to a file on the host
```
Comment (stdweird) [inline ncm-grub/src/main/pan/components/grub/schema.pan]: @ned21 added (but not tested)

File: ncm-grub/src/main/pan/components/grub/schema.pan [hunk 6]
```diff
@@ -16,37 +16,74 @@ include 'quattor/schema';
...
+    @{See description of the file option. The user (first field) to be picked from a password field.}
     "file_user" : string = "root"
 } with {
     if (is_defined(SELF["enabled"]) && SELF["enabled"]
         && !is_defined(SELF["file"]) && !is_defined(SELF["password"])) {
             error("specify either a hashed password or file to retrieve it from.");
     };
+    if (is_defined(SELF["file"]) && is_defined(SELF["password"])) {
+        error("specify either a hashed password or file to retrieve it from, not both.");
+    }
```
Comment (ned21) [inline ncm-grub/src/main/pan/components/grub/schema.pan]: Missing ;

**Comment by stdweird** [comment]
@ned21 thx, and fixed

**Comment by ned21** [comment]
👍 Are we OK to start merging 17.2 items?

**Comment by stdweird** [comment]
not yet i think, final 16.12 will be for later today i think

**Comment by jrha** [comment]
16.12.0 is built, feel free to merge.

**Pull request merged**